mitiq/lre/multivariate_scaling/layerwise_folding.py

```diff
@@ -0,0 +1,209 @@
+from mitiq.zne.scaling.folding import _check_foldable
+
+
+def _get_num_layers_without_measurements(input_circuit: Circuit) -> int:
+    """Checks if the circuit has non-terminal measurements and returns the
+    number of layers in the input circuit without the terminal measurements.
+
+        Args:
+            input_circuit: Circuit of interest
+
+        Returns:
+            num_layers: the number of layers in the input circuit without the
+                terminal measurements
+
+    """
+
+    _check_foldable(input_circuit)
+    circuit = deepcopy(input_circuit)
+    _pop_measurements(circuit)
+    return len(circuit)
+
+
+def _get_chunks(
+    input_circuit: Circuit, num_chunks: Optional[int] = None
+) -> List[Circuit]:
+    """Splits a circuit into approximately equal chunks.
+
+    Adapted from:
+    https://stackoverflow.com/questions/2130016/splitting-a-list-into-n-parts-of-approximately-equal-length
+
+        Args:
+            input_circuit: Circuit of interest
+            num_chunks: Number of desired approximately equal chunks
+                * when num_chunks == num_layers, the original circuit is
+                    returned
+                * when num_chunks == 1, the entire circuit is chunked into 1
+                    layer
+        Returns:
+            split_circuit: Circuit of interest split into approximately equal
+                chunks
+
+        Raises:
+            ValueError:
+                When the number of chunks for the input circuit is larger than
+                    the number of layers in the input circuit.
+
+            ValueError:
+                When the number of chunks is less than 1.
+
+    """
+    num_layers = _get_num_layers_without_measurements(input_circuit)
+    if num_chunks is None:
+        num_chunks = num_layers
+
+    if num_chunks < 1:
+        raise ValueError(
+            "Number of chunks should be greater than or equal to 1."
+        )
+
+    if num_chunks > num_layers:
+        raise ValueError(
+            f"Number of chunks {num_chunks} cannot be greater than the number"
+            f" of layers {num_layers}."
+        )
+
+    k, m = divmod(num_layers, num_chunks)
+    return [
+        input_circuit[i * k + min(i, m) : (i + 1) * k + min(i + 1, m)]
+        for i in range(num_chunks)
+    ]
+
+
+def _get_scale_factor_vectors(
+    input_circuit: Circuit,
+    degree: int,
+    fold_multiplier: int,
+    num_chunks: Optional[int] = None,
+) -> List[Tuple[Any, ...]]:
+    """Returns the patterned scale factor vectors required for multivariate
+    extrapolation.
+
+        Args:
+            input_circuit: Circuit to be scaled
+            degree: Degree of the multivariate polynomial
+            num_chunks: Number of desired approximately equal chunks
+            fold_multiplier: Scaling gap required by unitary folding
+
+        Returns:
+            scale_factor_vectors: Multiple variations of scale factors for each
```

Comment:
[nit]: The term "multiple variations" might be misleading. There need not be any variations (i.e. they could all be the same across the board for each layer). Perhaps it might be better described as "A vector of scale factors where each component in the vector corresponds to the layer in the input circuit."